user api

casperhart opened the pull request:
closes #92

I've made the changes to the user interface decided in #92, including:
- adding `detour` and `tour_path` functions
- `removing the `animate_tour` function
- updating examples and documentation, including vignettes and pkgdown site
- adding the `projection` aesthetic argument for the tour columns

The new interface can be used as follows:

```r
library(detourr)

detour(tourr::olive, tour_aes(projection = palmitic:eicosenoic, colour = region)) %>%
  tour_path(fps = 60) %>%
  display_scatter(alpha = 0.7)
```

casperhart commented:
I've since fixed the issue I was having with Shiny, so now everything should be updated to the new interface. The tests have been updated and all pass, and `devtools::check()` doesn't highlight any major issues.

earowang commented on DESCRIPTION:
```diff
@@ -20,12 +20,16 @@ Imports:
     viridisLite,
     grDevices,
     jsonlite,
+    magrittr,
```
I don't think you need to import {magrittr} now for the pipe, since base R introduced `|>` as a native pipe. In your vignettes and readme, you can use `|>` throughout.

earowang commented on R/detour.R:
```diff
@@ -0,0 +1,75 @@
+#' `tour_path` and `display_functions` is used to build a detour visualisation.
+#'
+#' @param .data a data frame, tibble, or crosstalk::SharedData object
+#' @param mapping a mapping of data columns to aesthetic values using the
+#' `tour_aes` function. The only required aesthetic is `projection`, which
+#' determines which columns are used to generate the tour path.
```
```suggestion
#' @param mapping a mapping of data columns to aesthetic values using the
#' `tour_aes` function. The only required aesthetic is `projection`, which
#' determines which columns are used to generate the tour path and supports tidy selection.
```

earowang commented on R/detour.R:
```diff
@@ -0,0 +1,75 @@
+    crosstalk_group <- .data$groupName()
+    .data <- .data$origData()
+    crosstalk_dependencies <- crosstalk::crosstalkLibs()
+  } else if (!inherits(.data, "data.frame")) {
```
```suggestion
  } else if (!is.data.frame(.data)) {
```

earowang commented on R/detour.R:
```diff
@@ -0,0 +1,75 @@
+    ))
+  }
+
+  structure(list(
+    mapping = mapping[!names(mapping) == "projection"],
+    config = NULL,
+    widget = NULL,
+    crosstalk = list(
+      crosstalkIndex = crosstalk_key,
+      crosstalkGroup = crosstalk_group
+    ),
+    projectionMatrices = NULL,
+    dataset = dataset,
+    crosstalk_dependencies = crosstalk_dependencies
+  ), class = "detour")
```
```suggestion
  structure(dataset,
    mapping = mapping[!names(mapping) == "projection"],
    projectionMatrices = NULL,
    config = NULL,
    widget = NULL,
    crosstalk = list(
      crosstalkIndex = crosstalk_key,
      crosstalkGroup = crosstalk_group
    ),
    crosstalk_dependencies = crosstalk_dependencies, 
    class = "detour")
```

`detour()` should return the `dataset` as the result (not the whole list) and others go to attributes. Attributes are metadata that users don't have to dive into details.

`tour_path()` should probably return `projectionMatrices` or `bases` as the result.

casperhart replied:
I would prefer that `detour()` and `tour_path` return similar objects. I.e. both the dataset and projection matrices, even if the projection matrices are null. All the rest can be attributes and that shouldn't be a problem.
However, `htmlwidgets::createWidget` expects a list. Would that just be a matter of creating an `as.list` method for the s3 object?

earowang replied:
Yes, they should return similar results, since it's a detour object. It's better to return projection matrices not the dataset. So `detour()` would return `NULL`.

Yes, `as.list()` for `detour`, ready for `createWidget()`.

earowang reviewed (commented):
When you define a new S3 class, please also define the `print()` method for it. In this case, `detour()` prints `dataset`, and `tour_path()` prints `projectionMatrices` or `bases`.

earowang commented on R/display_scatter.r:
```diff
@@ -45,87 +44,106 @@
+#'   tour_path(grand_tour(3), fps = 60) %>%
+#'   display_scatter(alpha = 0.7, axes = FALSE)
+#' @export
+display_scatter <- function(d,
```
```suggestion
display_scatter <- function(x,
```

Conventionally, use `x`.

casperhart replied:
I believe the convention is to use `p` for plots, which begs the question: would we want the `detour` object to be plottable from the start? I.e. have `detour` return an `htmlwidget` with a default tour path and display method?

earowang replied:
Either `x` or `p` works, I suppose. I'd like to have `tour_path()` to be traceable before plotting. `display_*()` makes the `detour` object plottable.

earowang commented on R/display_scatter.r:
```diff
@@ -45,87 +44,106 @@
-  }
-
+                            paused = TRUE,
+                            scale_factor = NULL) {
```
Abort if `x` is not a `detour` object.

casperhart commented:
Hi, I've addressed all of the comments so far, and implemented a very simple print method like so:

```
r$> print(dt)                                                                                                                                                                                                                                                              
Object of class `detour` with 10 bases and 493 animation frames
```

Let me know if there's any other info you would like to add to it. I would like it to behave similarly for both the result of `detour()` and for the result of `tour_path()`

casperhart commented:
Projection matrices in the detour object are currently stored as a list of matrices. It would be easy enough to store as a nested tbl with a column indicating whether the matrix in that row is a basis or is interpolated. This would make the printing a bit nicer, but would require tibble as a dependency. Thoughts?

earowang commented:
Is {detourr} already dependent on {tibble}? Yes, a nested tibble would be better.

earowang commented:
I think `print()` should print an empty/nested tibble instead of `cat()`, and at the end of `print()`, please return `invisible(x)`.

casperhart commented:
```r
library(detourr)
d = detour(tourr::flea, tour_aes(projection = -species, colour=species))
print(d)
#> # A tibble: 0 × 2
#> # … with 2 variables: is_new_basis <lgl>, projection_matrix <list>
d = d |> tour_path(grand_tour(3))
print(d)
#> # A tibble: 453 × 2
#>    is_new_basis projection_matrix
#>    <lgl>        <list>           
#>  1 TRUE         <dbl [6 × 3]>    
#>  2 FALSE        <dbl [6 × 3]>    
#>  3 FALSE        <dbl [6 × 3]>    
#>  4 FALSE        <dbl [6 × 3]>    
#>  5 FALSE        <dbl [6 × 3]>    
#>  6 FALSE        <dbl [6 × 3]>    
#>  7 FALSE        <dbl [6 × 3]>    
#>  8 FALSE        <dbl [6 × 3]>    
#>  9 FALSE        <dbl [6 × 3]>    
#> 10 FALSE        <dbl [6 × 3]>    
#> # … with 443 more rows
```

earowang commented on R/detour.R:
```diff
@@ -0,0 +1,108 @@
+    ))
+  }
+
+  x <- new_tibble(tibble(is_new_basis = logical(0), projection_matrix = list()),
```
```suggestion
new_tibble(tibble(is_new_basis = logical(0), projection_matrix = list()),
```
